### PR TITLE
chore(backend): use go build instead of bazel build for api server's docker file

### PR DESCRIPTION
**Pull request by jingzhang36**
**Description of your changes:**

Use go build instead of bazel build for api server's docker

Part of https://github.com/kubeflow/pipelines/issues/3250

**Checklist:**
- [ ] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 

   PR titles examples:
    * `fix(frontend): fixes empty page. Fixes #1234`
       Use `fix` to indicate that this PR fixes a bug.
    * `feat(backend): configurable service account. Fixes #1234, fixes #1235`
       Use `feat` to indicate that this PR adds a new feature. 
    * `chore: set up changelog generation tools`
       Use `chore` to indicate that this PR makes some changes that users don't need to know.
    * `test: fix CI failure. Part of #1234`
        Use `part of` to indicate that a PR is working on an issue, but shouldn't close the issue when merged.

- [ ] Do you want this pull request (PR) cherry-picked into the current release branch?
    
    If yes, use one of the following options:
  
    * **(Recommended.)** Ask the PR approver to add the `cherrypick-approved` label to this PR. The release manager adds this PR to the release branch in a batch update.
    *  After this PR is merged, create a cherry-pick PR to add these changes to the release branch. (For more information about creating a cherry-pick PR, see the [Kubeflow Pipelines release guide](https://github.com/kubeflow/pipelines/blob/master/RELEASE.md#option--git-cherry-pick).)

### CODE REVIEW

**Comment by Ark-kun** [comment]
How does this solution generate the code from .proto files?

**Comment by jingzhang36** [comment]
> How does this solution generate the code from .proto files?

Very good question. Let me show you my intended plan:

Current status:
[generate_api.sh](https://github.com/kubeflow/pipelines/blob/master/backend/api/generate_api.sh)  (which generates *.json, go_http_client and go_client) + Dockerfile (which builds upon the generated and the handwritten code) => api server image

After this PR:
generate_api.sh is still responsible for generating *.json, go_http_client and go_client. But replace the old Dockerfile with go-build based Dockerfile. And thus it is still generate_api.sh + Dockerfile => api server image

Last:
We totally remove generate_api.sh, and instead, in Dockerfile, we need to add commands like
```go get github.com/golang/protobuf/protoc-gen-go
protoc <a bunch of flags> <proto files>
....```
to generate *.json, go_http_client and go_client INSIDE the container. It is like we moving the operations in the  generate_api.sh to inside the container defined by our backend Dockerfile. That is feasible, because our current generate_api.sh is using protoc tools which is a go tool/lib after all (https://github.com/grpc-ecosystem/grpc-gateway). 
And we also get the bonus that we don't need to explicitly store the generated json,go_http_client and go_client in our repo any more. After all, those generated files are neither human-friendly to read nor editable by human. And, if some one really need to read the generated files (e.g., when we have to debug client related errors), you still can use the docker build --target mode to debug the intermediate stage (i.e., the generated files in the intermediate stage)

Since the last step's change is working on replacing generate_api.sh, so I separate it from this PR. Also, a smaller pr is easier to debug and verfiy.

File: backend/Dockerfile [hunk 1]
```diff
@@ -1,32 +1,15 @@
...
-RUN apt-get update && \
-  apt-get install -y cmake clang musl-dev openssl
+# 1. Build api server application
+# Use golang:1.13.1-stretch to keep GLIBC at 2.24 https://github.com/gotify/server/issues/225
+FROM golang:1.13.1-stretch as builder
+RUN apt-get update && apt-get install -y cmake clang musl-dev openssl
 WORKDIR /go/src/github.com/kubeflow/pipelines
-
 COPY . .
+RUN go mod vendor
```
Comment (Bobgy) [inline backend/Dockerfile:7]: Why is vendor required?
I thought we can just use go module directly

**Pull request merged**